crates/task-impls/src/quorum_proposal_recv.rs

```diff
@@ -0,0 +1,164 @@
+    pub private_key: <TYPES::SignatureKey as SignatureKey>::PrivateKey,
+
+    /// Reference to consensus. The replica will require a write lock on this.
+    pub consensus: Arc<RwLock<Consensus<TYPES>>>,
+
+    /// View number this view is executing in.
+    pub cur_view: TYPES::Time,
+
+    /// Network for all nodes
+    pub quorum_network: Arc<I::QuorumNetwork>,
+
+    /// Membership for Quorum Certs/votes
+    pub quorum_membership: Arc<TYPES::Membership>,
+
+    /// Membership for Timeout votes/certs
+    pub timeout_membership: Arc<TYPES::Membership>,
+
+    /// timeout task handle
+    pub timeout_task: Option<JoinHandle<()>>,
+
+    /// View timeout from config.
+    pub timeout: u64,
+
+    /// Output events to application
+    pub output_event_stream: async_broadcast::Sender<Event<TYPES>>,
+
+    /// This node's storage ref
+    pub storage: Arc<RwLock<I::Storage>>,
+
+    /// The most recent proposal we have, will correspond to the current view if Some()
+    /// Will be none if the view advanced through timeout/view_sync
+    pub current_proposal: Option<QuorumProposal<TYPES>>,
+
+    /// Spawned tasks related to a specific view, so we can cancel them when
+    /// they are stale
+    pub spawned_tasks: BTreeMap<TYPES::Time, Vec<JoinHandle<()>>>,
+
+    /// The node's id
+    pub id: u64,
+}
+
+#[cfg(feature = "dependency-tasks")]
+impl<TYPES: NodeType, I: NodeImplementation<TYPES>> QuorumProposalRecvTaskState<TYPES, I> {
+    /// Cancel all tasks that have been spawned before the provided view.
+    async fn cancel_tasks(&mut self, view: TYPES::Time) {
+        let keep = self.spawned_tasks.split_off(&view);
+        let mut cancel = Vec::new();
+        while let Some((_, tasks)) = self.spawned_tasks.pop_first() {
+            let mut to_cancel = tasks.into_iter().map(cancel_task).collect();
+            cancel.append(&mut to_cancel);
+        }
+        self.spawned_tasks = keep;
+        join_all(cancel).await;
+    }
+
+    /// Handles all consensus events relating to propose and vote-enabling events.
+    #[instrument(skip_all, fields(id = self.id, view = *self.cur_view), name = "Consensus replica task", level = "error")]
+    pub async fn handle(
+        &mut self,
+        event: Arc<HotShotEvent<TYPES>>,
+        event_stream: Sender<Arc<HotShotEvent<TYPES>>>,
+    ) {
+        match event.as_ref() {
+            HotShotEvent::QuorumProposalRecv(proposal, sender) => {
+                match handle_quorum_proposal_recv(proposal, sender, event_stream.clone(), self)
+                    .await
+                {
+                    Ok(Some(current_proposal)) => {
+                        self.current_proposal = Some(current_proposal);
+
+                        // Build the parent leaf since we didn't find it during the proposal check.
+                        let consensus = self.consensus.read().await;
+
+                        broadcast_event(
+                            Arc::new(HotShotEvent::VoteNow(
```

Comment:
Thinking about this more, I don't think we can send `VoteNow` here since `handle_quorum_proposal_recv` doesn't call `vote_if_able` to check the eligibility to vote.

Reply:
Isn't eligibility determined somewhat arbitrarily? It calls it only if we have a `current_proposal`, meaning that in this case where we receive an output from this function we should still be able to *try* to vote.